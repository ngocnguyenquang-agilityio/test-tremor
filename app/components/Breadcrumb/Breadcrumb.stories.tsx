import type { Meta, StoryObj } from "@storybook/react";

// Components
import Breadcrumb from "./Breadcrumb";

const meta = {
  title: "Components/Breadcrumb",
  component: Breadcrumb,
  tags: ["autodocs"],
} as Meta<typeof Breadcrumb>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Dashboard: Story = {
  render: () => <Breadcrumb />,
};
