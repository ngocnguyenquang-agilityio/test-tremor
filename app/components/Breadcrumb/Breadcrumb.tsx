// Libs
import { useParams } from "next/navigation";
import { memo } from "react";
import Link from "next/link";

// Components
import { Title } from "@tremor/react";
import { MdHome } from "react-icons/md";

// Constants
import { ORDER_LIST_REGEX, PRODUCT_LIST_REGEX, ROUTES } from "@/constants";

// Helpers
import { getCrumbName, isBrowser } from "@/helpers";

export interface BreadcrumbProps {
  isScrolled?: boolean;
  pathname: string;
}

const Breadcrumb = ({
  isScrolled = false,
  pathname,
}: BreadcrumbProps): JSX.Element => {
  const params = useParams();
  const isMobile = isBrowser && window.innerWidth <= 768;
  const isProjectPage = pathname === ROUTES.PROJECTS;
  const isStickyHeader = !isScrolled && isProjectPage;
  const newPath = pathname?.split("/").filter(path => path);

  const renderTitle = (path?: string) => {
    if (path) {
      if (path.match(ORDER_LIST_REGEX)) return "Order Details";
      if (path.match(PRODUCT_LIST_REGEX)) return "Product Details";
    }

    return newPath?.pop()?.replace("-", " ");
  };

  const renderCrumb = () => {
    return (
      <>
        {newPath?.map((link, index) => {
          const href = `/${newPath.slice(0, index + 1).join("/")}`;
          const crumbItemTextWhite = isProjectPage && "text-white";
          const activeCrumb =
            pathname === href
              ? "text-tremor-content-title dark:text-dark-tremor-content-title  "
              : "text-primary dark:text-dark-primary opacity-50  ";

          return (
            <>
              <li
                className={`flex gap-2 bc-link text-sm capitalize tracking-[0.02857em] leading-[0] ${crumbItemTextWhite} ${activeCrumb}`}>
                <Link href={href}>
                  {getCrumbName({
                    name: link,
                    path: href,
                    params: params?.id,
                  })}
                </Link>
                {newPath.length !== index + 1 && (
                  <span className={`${isProjectPage && "text-white"}`}>
                    &#47;
                  </span>
                )}
              </li>
            </>
          );
        })}
      </>
    );
  };

  return (
    <nav className={`${isProjectPage ? "pl-3 z-20" : isMobile ? "pl-3" : ""}`}>
      <ol className="flex flex-wrap gap-2 items-center text-gray-400">
        <li className="text-sm flex items-center gap-2 capitalize dark:text-dark-primary tracking-[0.02857em] leading-[0]">
          <Link href={ROUTES.HOME}>
            <MdHome className="text-lg" />
          </Link>
          &#47;
        </li>
        {renderCrumb()}
      </ol>
      <Title
        className={`text-tremor-content-title dark:text-dark-tremor-content-title font-bold capitalize tracking-[0.0075em] ${
          isStickyHeader ? "text-white" : "text-primary"
        }`}>
        {renderTitle(pathname)}
      </Title>
    </nav>
  );
};

export default memo(Breadcrumb);
