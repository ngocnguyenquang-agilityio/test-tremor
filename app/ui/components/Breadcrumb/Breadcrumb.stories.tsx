import type { Meta, StoryObj } from "@storybook/react";

// Components
import Breadcrumb from "./Breadcrumb";

const meta = {
  title: "Components/Common/Breadcrumb",
  component: Breadcrumb,
  tags: ["autodocs"],
  argTypes: {
    isScrolled: { description: "Scrolled for sticky header or not" },
    pathname: { description: "Path name of page" },
  },
} as Meta<typeof Breadcrumb>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Dashboard: Story = {
  args: {
    isScrolled: false,
    pathname: "/analytics",
  },
};
