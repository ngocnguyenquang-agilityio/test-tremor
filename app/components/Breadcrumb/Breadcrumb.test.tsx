import { render } from "@testing-library/react";

// Components
import Breadcrumb from "./Breadcrumb";

describe("Testing breadcrumb component", () => {
  it("should render correctly with display breadcrumb text", () => {
    const { container } = render(<Breadcrumb />);
    const link = container.getElementsByClassName("bc-link")[0];
    expect(link.textContent).toEqual("dashboards");
  });

  it("Should match snapshot", () => {
    const { container } = render(<Breadcrumb />);
    expect(container).toMatchSnapshot();
  });
});
