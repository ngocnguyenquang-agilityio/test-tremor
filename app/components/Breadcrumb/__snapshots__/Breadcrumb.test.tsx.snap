// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing breadcrumb component Should match snapshot 1`] = `
<div>
  <nav
    class=""
  >
    <ol
      class="flex flex-wrap gap-2 items-center text-gray-400"
    >
      <li
        class="text-sm flex items-center gap-2 capitalize dark:text-dark-primary tracking-[0.02857em] leading-[0]"
      >
        <a
          href="/dashboards/analytics"
        >
          <svg
            class="text-lg"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
            />
          </svg>
        </a>
        /
      </li>
      <li
        class="flex gap-2 bc-link false text-sm capitalize text-tremor-content-title dark:text-dark-tremor-content-title  tracking-[0.02857em] leading-[0]"
      >
        <a
          href="/analytics"
        >
          analytics
        </a>
      </li>
    </ol>
    <p
      class="text-tremor-title dark:text-dark-tremor-content-title font-bold capitalize tracking-[0.0075em] text-primary"
    >
      analytics
    </p>
  </nav>
</div>
`;
